### ISSUE TITLE
Add an option continue messages processing in batches right away.

### ISSUE THREAD

**Issue by lbilger** [holunda-io/camunda-bpm-correlate#40]
### Scenario

* Library version: 1.0.1
* Description of your use case: (detailed description or executable reproducer, e.g. GitHub repo)

Many processes contain asynchronous continuations in many places. If correlation of a message makes the process run through an async continuation, correlation of the next message in batch directly after the first will probably fail because the process is not yet ready to accept another message.

In this case it would be better not to batch at all and to process single messages in the order they come in. Or maybe even to prevent processing more than a single message per batch key in a single scheduler run.

### Current Behaviour

Messages are always processed in batches.

### Wanted Behaviour

Batch processing can be turned off.

### Possible Workarounds

It might be possible to prevent batching by fiddling around with the `correlationSelector`, e.g. adding a UUID to the `correlationVariables` of the `CorralationHint`. Haven't tried that, though.

**Comment by zambrovski** [comment]
Hi Lars,

Before building it, I'd like to consider several alternatives.

1. Introduce a switch to batch or not to batch
2. Introduce a batch size option.

Currently, I think the batch size is more generic. The correlation will select a batch containing only the top element. This means instead of introducing a no-batch option, we could introduce a batch-size option and set it to 1. By doing so we would achieve the same behaviour without changing the processing logic a lot.

What do you think?

Cheers,

Simon

**Comment by lbilger** [comment]
Hi Simon,

I guess the batch size option would be ok as well. The question is, is there any situation where batching will really bring an advantage? Leaving out effects like maybe database caching, it shouldn't make a performance difference if we correlate multiple related messages in a row or if we just correlate messages in the order they come in. So maybe things could be simplified by removing the batching altogether.

In our case, we could even use an _anti-batching_ option, meaning that two related messages should not be processed within a certain time-frame. But that would be a separate feature, I think.

The only thing I can imagine where batching really makes sense would be if the whole batch is passed to the application with the option to consolidate it, so that maybe only one message needs to be correlated instead of multiple messages (depends highly on application logic, maybe only the latest message is relevant or data from the messages can be combined).

Best
Lars

**Comment by zambrovski** [comment]
Hi Lars,

for the feature - will do it by limiting the batch size then.
for the anti-batching feature - just create an issue on that. This is then indeed something like setting a batch-delay.

For your understanding why batching is so important.
Imagine you get the messages:

A1, B1, A2, B2, C1, C2

with the letter denoting the target process instance (so messages with the same latter have the same correlation hint).

The correlator selects all messages and builds batches (groupings):

(A1, A2), (B1, B2), (C1, C2) and then tries to correlate. By introducing batch size, we would limit the batch AFTER batch creation resulting in
(A1), (B1), (C1) and then tries to correlate. Imagine the process instance A is not ready to receive the message yet, but B and C are. This first attempt would result in:

A1 -> fail (messages B1 and C1 will be successfully correlated).
In the next run it would create the following batches:

(A1 -> fail, A2), (B2), (C2), again reduced to (A1 -> fail), (B2), (C2) correlating B,C again would result in A1 -> fail/skipped (because it is not due yet).

It would continue to select the (A1, A2) batch and skip it until A1 next retry is due. Then it would correlate A1, if successful it would then get A2 and correlate that.


If you don't have batching at all, the grouping would not take place. This means that the initial messages will be interpreted as  A1, B1, A2, B2, C1, C2, resulting in first correlation attempt in A1 -> fail, B1 -> success, A2 -> fail, B2 -> success, C1 -> success, C2 -> success.

Now I hope this demonstrates the possible problem. A2 has been attempted to correlate after a failed correlation of A1. It could be ok, depending on your process, but could be a problem too, because you would violate the order of messages correlated with process A.

The preservation of order is one of the features we really want to support. If you don't want this, you can set error mode to continue, instead of first fail.

So if the order doesn't matter for you at all, you should not have any problems with the batch size anyways...


Hope this helps.

Cheers,

Simon







On Wed, 2023-01-18 at 00:12 -0800, lbilger wrote:

Hi Simon,

I guess the batch size option would be ok as well. The question is, is there any situation where batching will really bring an advantage? Leaving out effects like maybe database caching, it shouldn't make a performance difference if we correlate multiple related messages in a row or if we just correlate messages in the order they come in. So maybe things could be simplified by removing the batching altogether.

In our case, we could even use an anti-batching option, meaning that two related messages should not be processed within a certain time-frame. But that would be a separate feature, I think.

The only thing I can imagine where batching really makes sense would be if the whole batch is passed to the application with the option to consolidate it, so that maybe only one message needs to be correlated instead of multiple messages (depends highly on application logic, maybe only the latest message is relevant or data from the messages can be combined).

Best
Lars

**Comment by zambrovski** [comment]
Hi Lars,

I implemented the feature and wanted to try it out. I found out, that currently the example is not properly working. The problem with it is the de-serialization of the messages sent to
the reservation service. I seem to have broken the Kafka setup in Spring Cloud Streams and
get a message payload contained of one byte instead of the bytes representing my JSON.

Could you have a look on the configuration of the reservation service (especially the Spring Cloud Streams config for the binder)?

As soon we can validate, I would release the new version of the library.

Thanks,

Simon

**Comment by zambrovski** [comment]
@lbilger any chance to look on it this week?

**Comment by lbilger** [comment]
I will try to have a look tomorrow.

**Comment by lbilger** [comment]
The problem is not in the Spring Cloud configuration but in the shell script that sends the message. `kcat` splits the input into lines by default and creates a new message from each line.
It seems I can't commit to this repo and I don't want to fork for changing 2 lines, so I'll just post my suggestion here:
You can either use `jq -c` to output the message to a single line or use `kcat -D "\0"` to avoid splitting in kcat:

```shell
echo "$JSON" | "$CLIENT_BIN" \
  -D "\0"  \
  -b "$KAFKA_BOOTSTRAP_SERVER_HOST:$KAFKA_BOOTSTRAP_SERVER_PORT" \
  -t "$KAFKA_TOPIC_CORRELATE_FLIGHTS" -P
```

**Comment by zambrovski** [comment]
Thank you for helping with this. I'll close the issue as implemented and release the 1.1.0
